### PR TITLE
Rebase

### PR DESCRIPTION
**Pull request by ghabs**
Rebasing mainchainBMM with bitcoin:master.

A few notes for testing:
Most of src/test passed. The ones that didn't that I haven't been able to fix yet:
- script_tests.cpp : I reverted to the one currently in drivechain (and removed two from script_test.json CHECKLOCKTIMEVERIFY and CHECKSEQUENCEVERIFY)
- coins_tests.cpp : some type of error with the serialization test. I'm not sure if this is a "real" error, but I have commented it out for now.
I haven't tested the full workflow (mainchain->sidechain and sidechain->mainchain) from start to finish, but sidechaindb_tests.cpp passes.

PTAL!

### CODE REVIEW

**Comment by ghabs** [comment]
I noticed an error in this build - the sidechain addresses no longer autopopulate in the senddeposit.cpp dialog. I suspect that the error was caused by me moving the pwallet->AddWatchOnly to 
https://github.com/ghabs/bitcoin/blob/rebased/src/wallet/init.cpp#L273
as part of the refactor to use the vector of pwallets.

**Comment by CryptAxe** [comment]
@ghabs taking a look!

File: src/qt/sidechaindepositdialog.cpp [hunk 1]
```diff
@@ -80,7 +82,7 @@ void SidechainDepositDialog::on_pushButtonDeposit_clicked()
     const CAmount& nValue = ui->payAmount->value();
     CTransactionRef tx;
     std::string strFail = "";
-    if (!pwalletMain->CreateSidechainDeposit(tx, strFail, nSidechain, nValue, keyID)) {
+    if (!vpwallets[0]->CreateSidechainDeposit(tx, strFail, nSidechain, nValue, keyID)) {
```
Comment (CryptAxe) [inline src/qt/sidechaindepositdialog.cpp]: I think we need ```if (!vpwallets.empty())``` first here, I'm new to the multi wallet support code though so we should double check.

Comment (CryptAxe) [reply src/qt/sidechaindepositdialog.cpp]: Same comment for other lines and from what I can tell everywhere in the upstream codebase is checking that the vector of wallets isn't empty before accessing. One of us can add those checks before we merge this :+1:

Comment (ghabs) [reply src/qt/sidechaindepositdialog.cpp]: Good point - I've added isempty checks for vpwallets

**Comment by CryptAxe** [comment]
@ghabs thank you! Do you happen to know if there is any method of determining the currently selected wallet instead of always using the first created wallet (for sidechain deposits etc)? I will look into this as well

**Comment by ghabs** [comment]
Yeah not sure how, but you're definitely correct that my naive implementation using vpwallets[0] should be changed to enable deposits from the active wallet. I see that there is a setCurrentWallet func, but not how to get active wallet.

**Comment by ghabs** [comment]
Unrelated to the rebase but I think there might be a bug in the CCoinsViewCache::HaveInputs logic. It accepts a boolean value fSidechainInputs, and if it fSidechainInputs == true then it loops  through tx inputs, checks if one of the keys is the a sidechain key, and if so returns fSidechainInputs as true.

I suspect the if statement (line 270 in coins.cpp) should be if fSidechainInputs == NULL, not just if it is true, as in validation.cpp line 2022 fSidechainInputs is input as false and thus that logic won't ever execute.

If that's true I'll make that change in this commit.

**Comment by CryptAxe** [comment]
> Yeah not sure how, but you're definitely correct that my naive implementation using vpwallets[0] should be changed to enable deposits from the active wallet. I see that there is a setCurrentWallet func, but not how to get active wallet.

I'll look into it, might not have been added even upstream yet.

> 
> Unrelated to the rebase but I think there might be a bug in the CCoinsViewCache::HaveInputs logic. It accepts a boolean value fSidechainInputs, and if it fSidechainInputs == true then it loops through tx inputs, checks if one of the keys is the a sidechain key, and if so returns fSidechainInputs as true.
> 
> I suspect the if statement (line 270 in coins.cpp) should be if fSidechainInputs == NULL, not just if it is true, as in validation.cpp line 2022 fSidechainInputs is input as false and thus that logic won't ever execute.
> 

That's actually intentional but since it is unclear it should be refactored... maybe one bool and one bool by reference as parameters instead of trying to get too much out of the one pointer. Thanks for pointing that out, I think my code here could be made more clear.

**Comment by CryptAxe** [comment]
Also I'm going to try and fix the sidechain address watching issue @ghabs and I'll try to open a pull request into your branch some time today!

**Comment by ghabs** [comment]
Excellent thanks @CryptAxe !

**Comment by ghabs** [comment]
Multi-wallet isn't currently supported in the GUI
https://bitcoin.org/en/release/v0.15.0.1#multi-wallet-support
https://github.com/bitcoin/bitcoin/blob/3c098a8aa0780009c11b66b1a5d488a928629ebf/src/qt/bitcoin.cpp#L487

We could refactor the methods (ex listsidechaindeposits) to optionally take in a specified wallet, but that might be better as another issue rather than as part of this rebase.

**Comment by ghabs** [comment]
Re: the coins_tests.cpp failing I see that the failing test I commented out was actually the same one @CryptAxe commented out dealing with serialization
https://github.com/drivechain-project/bitcoin/blob/mainchainBMM/src/test/coins_tests.cpp#L433 

In hindsight this makes 100% sense as the addition of the fCriticalData flag would throw off the expected serialization. 
For now I left the serialize / unserialize function unchanged with the rebase, but I do have a version on a separate branch that contains an updated set of serialization functions and will (hopefully over weekend) have updated coins_tests to match
https://github.com/ghabs/bitcoin/blob/integrationtests/src/coins.h#L124

**Comment by CryptAxe** [comment]
@ghabs sorry I haven't opened a PR into your branch yet, I was trying to test it out and found this issue (thanks to checking this out) https://github.com/bitcoin/bitcoin/issues/12047

**Comment by ghabs** [comment]
Np! Yeah that's a weird bug, unable to replicate on my Ubuntu box

**Comment by ghabs** [comment]
sidechain address watch is working - in bitcoin-qt it comes up under the "receiving addresses"

**Comment by CryptAxe** [comment]
> sidechain address watch is working - in bitcoin-qt it comes up under the "receiving addresses"

@ghabs did you change anything?

I'm opening a PR right now with some updates for your branch. I'm going to change some things with ```CCoins``` -> ```Coins``` as well and I'll open up a PR for that when it's ready.

In regards to the multi wallet support issues, we'll just wait for upstream to add multiwallet support before we worry about using anything besides the first active wallet. If they don't add it in core in the next few months then we should try and add it ourselves and send it back to them!

**Comment by ghabs** [comment]
No it was a mistake on my part - I thought the watched addresses would appear on the sidechaindeposit dialog but forgot what it was actually supposed to do.
Thanks for the PR!

**Review by CryptAxe — APPROVED** [review]

**Comment by CryptAxe** [comment]
Ready to go @psztorc

**Pull request merged**